qclib/unitary.py

```diff
@@ -87,12 +87,26 @@ def _qsd(gate1, gate2):
     circuit = circuit.compose(right_gate, qubits[0:-1])
     return circuit
 
+def _unitary_condition(matrix):
+    return np.conj(matrix.T).dot(matrix)
+
+def _is_unitary(matrix):
+    return np.allclose(_unitary_condition(matrix), np.identity(matrix.shape[0]))
+
+def _closest_unitary(matrix):
+    svd_u,_,svd_v = np.linalg.svd(matrix)
+    return svd_u.dot(svd_v)
 
 def _compute_gates(gate1, gate2):
 
     d_square, gate_v = np.linalg.eig(gate1 @ gate2.conj().T)
-    list_d = np.sqrt(d_square)
+    list_d = np.sqrt(d_square, dtype=np.complex)
```

Comment:
I am not sure if we need to include `dtype=np.complex` . If d[k] is real, d[k]**2 >= 0 and if d[k] is complex we don't need to set the type.

Reply:
The matrix **gate1 @ gate2.conj().T** is unitary, so the only guarantee we have about the eigenvalues is that their absolute value is 1. There is no restriction on the phase.